bitcoin/src/blockdata/transaction.rs

```diff
@@ -833,6 +834,124 @@ impl Transaction {
     pub fn script_pubkey_lens(&self) -> impl Iterator<Item = usize> + '_ {
         self.output.iter().map(|txout| txout.script_pubkey.len())
     }
+
+    // TODO: Find a better reference for how taproot sigops are limited.
+    /// Counts the total number of sigops.
+    ///
+    /// This value is for pre-taproot transactions only.
+    ///
+    /// > In taproot, a different mechanism is used. Instead of having a global per-block limit,
+    /// > there is a per-transaction-input limit, proportional to the size of that input.
+    /// > ref: <https://bitcoin.stackexchange.com/questions/117356/what-is-sigop-signature-operation#117359>
+    ///
+    /// The `spent` parameter is an optional closure/function that takes in an [`OutPoint`] and
+    /// returns a [`TxOut`]. Without access to the previous [`TxOut`], any sigops in redeemScripts,
+    /// witnessScripts, and P2WPKH sigops will not be counted.
+    pub fn total_sigop_cost<S, F>(&self, mut spent: Option<S>) -> usize
+    where
+        S: FnMut(&OutPoint) -> Option<TxOut>,
+    {
+        // TODO: Use checked multiplication here and below.
+        let mut cost = self.count_p2pk_p2pkh_sigops() * 4;
+
+        // coinbase tx is correctly handled because `spent` will always returns None.
+        cost += self.count_p2sh_sigops(spent.as_mut()) * 4;
```

Comment:
There are two paths here I can see:

* Moving to `u64` everywhere, which kinda sucks because all the stdlib `.len()` functions return a usize and there is no `u64::from(usize)`. Forcing us to cast or panic.
* Circle back to `Script`/`ScriptBuf` and refuse to allow construction of scripts that exceed 2^31 bytes (say) so we can stop worrying about this.

In practice no script can exceed 4Mb because that's the maximum number of bytes in a block.

Reply:
Currently, the biggest contributor to sigops counts for transactions on mainnet is bare multisig in the output. (20 sigops hard coded per OP_CHECKMULTISIG)

Just counting those, it would require 53687092 bare OP_CHECKMULTISIGs in the outputs in order to overflow 30 bits.

With current consensus rules, we are nowhere close to a scenario where this will overflow for 32 bits.

u32 is more than enough. The decision on u32/u64/usize should be dictated solely by ergonomics of use.

At this point I think usize is the most ergonomic.

Reply:
While working on the weight/size stuff I stumbled across https://github.com/rust-bitcoin/rust-bitcoin/issues/843

We should be using saturating add/mul.